sway/tree/view.c

```diff
@@ -636,13 +636,51 @@ static void handle_foreign_activate_request(
 	}
 }
 
+static void handle_foreign_fullscreen_request(
+		struct wl_listener *listener, void *data) {
+	struct sway_view *view = wl_container_of(
+			listener, view, foreign_fullscreen_request);
+	struct wlr_foreign_toplevel_handle_v1_fullscreen_event *event = data;
+
+	// Match fullscreen command behavior for scratchpad hidden views
+	struct sway_container *container = view->container;
+	if (!container->workspace) {
+		while (container->parent) {
+			container = container->parent;
+		}
+	}
+
+	container_set_fullscreen(container,
+		event->fullscreen ? FULLSCREEN_WORKSPACE : FULLSCREEN_NONE);
+	if (event->fullscreen) {
+		arrange_root();
+	} else {
+		if (container->parent) {
+			arrange_container(container->parent);
+		} else if (container->workspace) {
+			arrange_workspace(container->workspace);
+		}
+	}
+}
+
 static void handle_foreign_close_request(
 		struct wl_listener *listener, void *data) {
 	struct sway_view *view = wl_container_of(
 			listener, view, foreign_close_request);
 	view_close(view);
 }
 
+static void handle_foreign_destroy(
+		struct wl_listener *listener, void *data) {
+	struct sway_view *view = wl_container_of(
+			listener, view, foreign_destroy);
+
+	wl_list_remove(&view->foreign_activate_request.link);
+	wl_list_remove(&view->foreign_fullscreen_request.link);
+	wl_list_remove(&view->foreign_close_request.link);
+	wl_list_remove(&listener->link);
```

Comment:
Nit: can we remove `&view->foreign_destroy.link` instead? Makes it easier to compare it with the function registering the listeners and check we aren't missing anything.